tools/deploy_firmware.py

```diff
@@ -0,0 +1,287 @@
+
+import time
+import sys
+import os
+import subprocess
+import difflib
+import argparse
+import shutil
+from typing import List
+
+import pexpect
+import pexpect.fdpexpect
+import pexpect.pxssh
+import serial
+
+
+class PrplwrtDevice:
+    """Represents a prplWrt device.
+
+    Offers methods to check if a device needs to be upgraded and to do the actual upgrade.
+
+    It needs to have access to the artifacts of a build job to determine when an upgrade
+    is needed (see `artifacts_dir`).
+    """
+
+    serial_prompt = "_pexpect_prompt_ "
+    """For serial connections we will set this prompt to make it easier to "expect" it."""
+
+    baudrate = 115200
+    """The baudrate of the serial connection to the device."""
+
+    def __init__(self, device: str, name: str, image: str, username: str = "root"):
+        """
+
+        Parameters
+        -----------
+        device: str
+            The name of the platform (example: netgear-rax40).
+        name: str
+            The name of the device (it should ne reachable through ssh without a password).
+        image: str
+            The name of the image that can be used to upgrade the device.
+        username: str, optional
+            The username to use when connecting to the device over SSH.
+        """
+        self.device = device
+        self.name = name
+        self.image = image
+        self.username = username
+
+        self.rootdir = os.path.join(os.path.dirname(os.path.realpath(__file__)), "..")
+        self.artifacts_dir = os.path.join(self.rootdir, "build/{}".format(self.device))
+        """The directory where artifacts are stored. It's expected to contain
+        prplwrt-version, and the image file."""
+
+    def set_prompt(self, pexp: pexpect):
+        """Set the prompt to `serial_prompt` on a pexpect object."""
+        pexp.sendline("export PS1='{}'".format(self.serial_prompt))
+
+    def read_remote_version(self) -> List[str]:
+        """Read prplwrt-version on a remote device
+
+        Returns
+        -------
+        List[str]
+            The content of the prplwrt-version file on the device as a list of strings.
+        """
+        version = None
+        with pexpect.pxssh.pxssh(logfile=sys.stdout.buffer) as shell:
+            shell.login(self.name, self.username)
+            shell.sendline("cat /etc/prplwrt-version")
+            shell.expect("cat /etc/prplwrt-version")
+            shell.prompt()
+            version = shell.before.decode().splitlines()
+        return version
+
+    def read_artifacts_dir_version(self) -> List[str]:
+        """Reads the local prplwrt-version.
+
+        The version file is read from the artifacts (see `artifacts_dir`).
+
+        Returns
+        -------
+        List[str]
+            The content of the local prplwrt-version as a list of strings.
+        """
+        with open(os.path.join(self.artifacts_dir, "prplwrt-version")) as version_file:
+            return version_file.read().splitlines()
```

Comment:
Might be nice in the future to store or return it as a map, e.g.
```python
return dict([(k,v) for k, v in (l.split('=') for l in version_file.read().splitlines())])
```

Reply:
That's what I initially wanted to do, but the content of the file might change a lot very soon (with the new prplwrt release candidate), and it could really be any format.
Since we really only need to know if there is any difference and not what the difference is, I just went with a list of lines instead.